libsForQt5.qttools: Add option for building qdoc

OPNA2608 opened the pull request:
###### Description of changes

Mirroring what's been done on qt6Packages.qttools, hide this behind a `withClang` option.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
  Tested new `qdoc` binary.
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

wineee commented on pkgs/development/libraries/qt-5/modules/qttools.nix:
```diff
@@ -11,6 +23,12 @@ qtModule {
     sed -i "src/qtattributionsscanner/qtattributionsscanner.pro" \
         -e '/^cmake_qattributionsscanner_config_version_file.input =/ s|$$\[QT_HOST_DATA.*\]|${lib.getDev qtbase}|'
+    sed -i "src/qdoc/qdoc.pro" \
+        -e '/^cmake_qdoc_config_version_file.input =/ s|$$\[QT_HOST_DATA.*\]|${lib.getDev qtbase}|'
```
why this is not part of `lib.optionalString withClang`  so we can rebuild 0 ?

OPNA2608 replied:
Low rebuild amount isn't really a concern for me with this PR. We haven't had Qt5 `qdoc` in a long time, I don't need this available ASAP, and none of my Lomiri packages have been written with a functioning `qdoc` in mind anyway.

In other packages, hiding non-conditional patches behind conditionals is usually frowned upon AFAICT. Increases risk of bitrotting when the patching only happens *sometimes*, or even *rarely*.

OPNA2608 replied:
Moved to a patch where we `substituteAll` the `lib.getDev qtbase` reference.

OPNA2608 commented:
Rebased on staging to resolve merge conflict.

OPNA2608 commented on pkgs/development/libraries/qt-5/modules/qttools.nix:
```diff
@@ -1,8 +1,20 @@
 
 qtModule {
   pname = "qttools";
   propagatedBuildInputs = [ qtbase qtdeclarative ];
+  buildInputs = lib.optionals withClang [
+    llvmPackages.libclang
+    llvmPackages.libllvm
+  ];
   outputs = [ "out" "dev" "bin" ];
```
Why not do that in a separate PR, across all of the Qt5 modules at once, where you wouldn't have to deal with it having to go through staging? I don't disagree with fixing the order, but this feels like the wrong PR to chain it to :confused:.

SuperSandro2000 replied:
> Why not do that in a separate PR, across all of the Qt5 modules at once, where you wouldn't have to deal with it having to go through staging?

That's way more work and will take way more time and will probably discover other issues. If we just pretend we didn't look at all the other construction sites and just fix this one, we can feel good about ourselves making one little piece of the code a bit better without starting the big work.

OPNA2608 committed:
libsForQt5.qttools: Move QT_HOST_DATA from sed calls to a patch

OPNA2608 committed:
libsForQt5.qttools: nixfmt

OPNA2608 committed:
libsForQt5.qttools: Reorder outputs vs propagatedBuildInputs

We usually order them the other way around.

OPNA2608 commented:
(Hopefully) not stale anymore, still interested in getting this in *eventually*.

- Moved the existing `sed` stuff into a patch as requested
- `nixfmt` while we're touching the file, just to reduce potential future nitpicking abit
- Reordered `outputs`, `buildInputs` and `propagatedBuildInputs` as requested, just to get that complaint off the table as well (though I still fail to see the point in insisting on that, since this is really just asking for an entire reorder of the file now…)
- Added `withClang` option to allow opting into `qdoc` building

K900 commented:
Why do we need this again, except "Qt6 has it"?

OPNA2608 commented:
Building QDocs on Qt5:

https://github.com/NixOS/nixpkgs/blob/fa12935a32af71ab3a8b4ff60ab3c37490eea336/pkgs/development/libraries/accounts-qml-module/default.nix#L61
https://github.com/NixOS/nixpkgs/blob/fa12935a32af71ab3a8b4ff60ab3c37490eea336/pkgs/desktops/lomiri/qml/lomiri-action-api/default.nix#L60
https://github.com/NixOS/nixpkgs/blob/fa12935a32af71ab3a8b4ff60ab3c37490eea336/pkgs/desktops/lomiri/services/lomiri-indicator-network/default.nix#L101
https://github.com/NixOS/nixpkgs/blob/fa12935a32af71ab3a8b4ff60ab3c37490eea336/pkgs/desktops/lomiri/services/lomiri-download-manager/default.nix#L12
https://github.com/NixOS/nixpkgs/blob/fa12935a32af71ab3a8b4ff60ab3c37490eea336/pkgs/desktops/lomiri/qml/lomiri-ui-toolkit/default.nix#L136
etc

SuperSandro2000 commented on pkgs/development/libraries/qt-5/modules/qttools.nix:
```diff
@@ -5,6 +5,9 @@
   qtdeclarative,
   substituteAll,
+  # clang-based c++ parser for qdoc and lupdate
+  withClang ? false,
```
Hmm, I am not so super happy with this option name. Maybe we should name it instead withQDoc or withTools or withExtraTools? That it required clang is some detail which isn't really exposing what we want to do.

OPNA2608 replied:
Hmm. I would be against that because Clang also enables https://wiki.qt.io/Lupdate_clang-based_c++_parser as the comment notes… but it looks like that functionality was actually never ported to Qt5.

Does it make sense to nonetheless keep the name the same, for consistency with `qt6.qttools`? Then I'll only adjust the comment to only mention qdoc.

https://github.com/NixOS/nixpkgs/blob/cb9a96f23c491c081b38eab96d22fa958043c9fa/pkgs/development/libraries/qt-6/modules/qttools.nix#L8-L9

SuperSandro2000 replied:
I am not sure, someone please decide 😓

OPNA2608 replied:
Option has been removed, we always build with clang stuff added now.

K900 commented:
Can we always build with clang, and then just move qdoc to the devtools output? Also, do we even care about these docs? Does _anyone_ use them?

wineee commented:
> Can we always build with clang, and then just move qdoc to the devtools output? 

May introduce unnecessary run-time dependencies for some applications [eg.](https://github.com/NixOS/nixpkgs/issues/239487),
Even if move qdoc to the devtools , lupdate will still depend on libclang because the clang-based C++ parser is enabled.

K900 commented:
lupdate should be in .bin, which should not be depended on by other things, generally.

OPNA2608 commented:
> Can we always build with clang, and then just move qdoc to the devtools output?

`qdoc` and `lupdate` are already in the `dev` output because they're build-time tools.

> May introduce unnecessary run-time dependencies for some applications [eg.](https://github.com/NixOS/nixpkgs/issues/239487),
Even if move qdoc to the devtools , lupdate will still depend on libclang because the clang-based C++ parser is enabled.

The `bin` output (which on Qt5 has the `libqquickwidget.so` the linked issue mentions, whether we should expose it like that is another issue I guess) has no dependency on the `dev` output or LLVM. I think this was only an issue on Qt6, where `qttools` isn't split into outputs anymore.

Example, by enabling `withClang` and looking at a package with `qttools` in `buildInputs` (not sure if that's really correct, but again besides the point):
![image](https://github.com/user-attachments/assets/ca7790e9-34b7-4f78-a98b-48cbb312bbdc)

So is it fine to just enable this?

K900 commented:
Yes, I think this is fine to enable unconditionally then, and maybe we should look into doing the same (i.e. split outputs + unconditional clang support) on 6.

K900 commented on pkgs/development/libraries/qt-5/modules/qttools.nix:
```diff
@@ -29,6 +35,12 @@ qtModule {
+  postPatch = ''
+    # llvm-config --includedir gives libllvm includedir, looks for libclang header there
+    substituteInPlace src/qdoc/configure.pri \
+      --replace-fail "\''$\''$CLANG_INCLUDEPATH/clang-c" "${lib.getDev llvmPackages.libclang}/include/clang-c"
```
Feels a bit weird to move sed calls to a patch and then add another sed call?

OPNA2608 replied:
Also moved to a patch now.

K900 replied:
Yeah, I definitely prefer that.

OPNA2608 committed:
libsForQt5.qttools: Add LLVM for qdoc

Since our Qt5 packaging uses outputs, we don't need to hide this behind an option like with Qt6.
